system/jlib/jstring.cpp

```diff
@@ -937,28 +937,34 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
 // Copy source to result, replacing all occurrences of "oldStr" with "newStr"
 StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char *source, size_t lenOldStr, const char* oldStr, size_t lenNewStr, const char* newStr)
 {
-    if (lenSource)
-    {
-        size_t left = lenSource;
-        while (left >= lenOldStr)
+    if (lenSource >= lenOldStr)
+    {
+        // Reserve space when the new string is longer than the old string
+        unsigned steps = lenSource - lenOldStr + 1;
+        unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
+        result.ensureCapacity(maxResultLength);
+
+        size_t offset = 0;
+        size_t lastCopied = 0;
+        size_t maxOffset = lenSource - lenOldStr;
+        char firstChar = oldStr[0];
+        while (offset < maxOffset)
         {
-            if (memcmp(source, oldStr, lenOldStr)==0)
+            if (unlikely(source[offset] == firstChar) && unlikely((lenOldStr == 1) || memcmp(source + offset, oldStr, lenOldStr)==0))
```

Comment:
style preference.  I would prefer this split over two lines to make it easier to read.  i.e. check first character, and if that matches compare the rest.

Reply:
Changed to 
```
 if (unlikely(source[offset] == firstChar) 
    && unlikely((lenOldStr == 1) || memcmp(source + offset, oldStr, lenOldStr)==0))
````

I'm not sure if this is what you meant, but adding an additional if statement would mean I also need another else incrementing output

```
 if (unlikely(source[offset] == firstChar) 
{     
    if (unlikely((lenOldStr == 1) || memcmp(source + offset, oldStr, lenOldStr)==0))
    {
        ...
    }
    else
         offset++;
 }
 else
     offset++;
```